### PR TITLE
Support for weighted dataset mixtures

### PR DESCRIPTION
**Pull request by akuramshin**
Created a new `MixtureDataset` Dataset type that supports weighted sampling from a list of datasets. Inspired by the [octo](https://github.com/octo-models/octo/tree/main) codebase that uses TensorFlow's [sample_from_datasets](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#sample_from_datasets) to create oxe mixtures.

Since torch does not have a `sample_from_datasets` equivalent, I recreate a minimal implementation using numpy's [RandomState](https://numpy.org/doc/1.16/reference/generated/numpy.random.RandomState.html) to ensure a reproducible sampling order based on the seed.

I added tests in `data_loader_test.py`, but let me know if these are insufficient and more should be added.

I also added an `accumulation_steps` parameter to the `AdamW` optimizer factory to support gradient accumulation. I saw [here](https://github.com/Physical-Intelligence/openpi/issues/286#issuecomment-2655112920) that you guys had plans to add grad accumulation, but I was not able to find it. Let me know if this should not be part of this PR.

### CODE REVIEW

**Review by uzhilinsky — COMMENTED** [review]
Huge thanks for the contribution. Adding a suggestion with the hope of making this easier to understand.

File: src/openpi/training/config.py [hunk 1]
```diff
@@ -90,6 +90,27 @@ class DataConfig:
     # If true, will disable syncing the dataset from the Hugging Face Hub. Allows training on local-only datasets.
     local_files_only: bool = False
 
+    mixture_configs: Sequence["DataConfig"] | None = None
```
[the remark this answers is not in the document]
Comment (akuramshin) [reply src/openpi/training/config.py:93]: Good idea! Let me first see if I am understanding your suggestion correctly:

We create `DataMixtureConfig` that inherits from `DataConfig` (we could also create a `DataConfigBase` Protocol since the MixtureConfig doesn't need the transform fields).

Then in `data_loader.py` we handle the mixture case in the `create_dataset()` and `transform_dataset()` functions via something like `if isinstance(data_config, _config.DataMixtureConfig)...`. This way the `create_data_loader()` function remains the same. (Just a small thing but when creating the `MixtureDataset` I set the seed from the TrainConfig, so I might have to pass that into `create_dataset`)

Also, should I create a `BaseDataConfigFactory(abc.ABC)` class to separate `DataConfigFactory` and `DataMixtureConfigFactory`?

Let me know if I am not understanding your point!

File: src/openpi/training/config.py [hunk 2]
```diff
@@ -90,6 +90,27 @@ class DataConfig:
...
+    mixture_weights: Sequence[float] | None = None
+    mixture_stop_on_empty: bool = False
+
+    def __post_init__(self):
+        if self.mixture_configs is None:
+            object.__setattr__(self, "mixture_configs", ())
```
[the remark this answers is not in the document]
Comment (akuramshin) [reply src/openpi/training/config.py:99]: Ah no good point this will be the cleaner approach

File: scripts/compute_norm_stats.py [hunk 3]
```diff
@@ -24,6 +24,24 @@ def create_dataset(config: _config.TrainConfig) -> tuple[_config.DataConfig, _da
     data_config = config.data.create(config.assets_dirs, config.model)
     if data_config.repo_id is None:
         raise ValueError("Data config must have a repo_id")
+    if data_config.mixture_configs:
+        datasets = [_data_loader.create_dataset(cfg, config.model) for cfg in data_config.mixture_configs]
+
+        transformed_datasets = []
+        for i, dataset in enumerate(datasets):
+            cfg = data_config.mixture_configs[i]
+            transformed_dataset = _data_loader.TransformedDataset(
```
[the remark this answers is not in the document]
Comment (akuramshin) [reply scripts/compute_norm_stats.py]: Extracted a `transform_dataset(dataset, data_config)` function

File: src/openpi/training/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (akuramshin) [reply src/openpi/training/config.py]: > Good idea! Let me first see if I am understanding your suggestion correctly:
> 
> We create `DataMixtureConfig` that inherits from `DataConfig` (we could also create a `DataConfigBase` Protocol since the `DataMixtureConfig` doesn't need the transform fields).
> 
> Then in `data_loader.py` we handle the mixture case in the `create_dataset()` and `transform_dataset()` functions via something like `if isinstance(data_config, _config.DataMixtureConfig)...`. This way the `create_data_loader()` function remains the same. (Just a small thing but when creating the `MixtureDataset` I set the seed from the TrainConfig, so I might have to pass that into `create_dataset`)
> 
> Also, should I create a `BaseDataConfigFactory(abc.ABC)` class to separate `DataConfigFactory` and `DataMixtureConfigFactory`?
> 
> Let me know if I am not understanding your point!

I completed a bit of a refactor according to my comments here. Let me know if this is not what you are looking for and I can refactor again @uzhilinsky
